### PR TITLE
Fix/23003: Placeholders are overlapping with auto-filled text

### PR DESCRIPTION
**Pull request by dukenv0307**
### Details
Placeholders are overlapping with auto-filled text. We need to update the state of label after running animation to fix that

### Fixed Issues

$ https://github.com/Expensify/App/issues/23003
PROPOSAL: https://github.com/Expensify/App/issues/23003#issuecomment-1643778453

### Tests

1. Make sure you are already logged out before visiting this link, if not reload the page after logging out.
2. Let the Login details be auto-filled (make sure not to click anywhere on the page before performing the next step)
3. Click on the share button on the address bar
4. Verify that the placeholder should not overlap auto-filled text

- [x] Verify that no errors appear in the JS console

### Offline tests

NA
### QA Steps

1. Make sure you are already logged out before visiting [this](https://staging.new.expensify.com/) link, if not reload the page after logging out.
2. Let the Login details be auto-filled (make sure not to click anywhere on the page before performing the next step)
3. Click on the share button on the address bar
4. Verify that the placeholder should not overlap auto-filled text

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/129500732/494e7edb-4eda-4742-9254-a745ea197acb

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/129500732/b1a65d4e-9893-40e1-b061-4879490fe3e0

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/129500732/a11f97d5-6d6a-4367-8e8e-9207051f3271

</details>

<details>
<summary>Desktop</summary>

This PR is only visible on web

</details>

<details>
<summary>iOS</summary>
This PR is only visible on web

</details>

<details>
<summary>Android</summary>
This PR is only visible on web

</details>

### CODE REVIEW

**Comment by situchan** [comment]
@dukenv0307 please pull main

File: src/components/TextInput/BaseTextInput.js [hunk 1]
```diff
@@ -185,17 +186,30 @@ function BaseTextInput(props) {
...
-    useEffect(() => {
+    const updateLabel = useCallback(() => {
         // We can't use inputValue here directly, as it might contain
         // the defaultValue, which doesn't get updated when the text changes.
         // We can't use props.value either, as it might be undefined.
-        if (hasValueRef.current || isFocused) {
+        if (hasValueRef.current || isFocused || isInputAutoFilled(input.current)) {
             activateLabel();
-        } else if (!hasValueRef.current && !isFocused) {
+        } else if (!hasValueRef.current && !isFocused && !isInputAutoFilled(input.current)) {
```
Comment (situchan) [inline src/components/TextInput/BaseTextInput.js:195]:
```suggestion
        } else {
```

File: src/components/TextInput/BaseTextInput.js [hunk 2]
```diff
@@ -185,17 +186,30 @@ function BaseTextInput(props) {
...
+        // This also avoids a bug where we immediately clear server errors when the loading indicator unmounts and Form remounts with server errors.
+        if (isFirstRender.current) {
+            isFirstRender.current = false;
+            InteractionManager.runAfterInteractions(() => {
+                updateLabel();
+            });
+            return;
+        }
+        updateLabel();
+    }, [updateLabel, inputValue]);
```
Comment (situchan) [inline src/components/TextInput/BaseTextInput.js:212]: Why `inputValue` is needed in dependency?

**Comment by situchan** [comment]
Found regression: label is not activated first time even though focused (for test, you should not have any autofill on browser or enable `Use your screen lock when filling passwords` in chrome password setting)


https://github.com/Expensify/App/assets/108292595/73734431-e737-4d58-909d-fe71f8c297b8

**Comment by situchan** [comment]
@dukenv0307 are you investigating?
Here's root cause:
https://stackoverflow.com/questions/55198517/react-usestate-why-settimeout-function-does-not-have-latest-state-value

**Comment by situchan** [comment]
> So I suggest that we should always use runAfterInteractions instead of only the first time


yes but as this is general component, it will apply to all inputs throughout the app where runAfterInteractions is not needed at all

**Comment by situchan** [comment]
> @situchan We can use accessibilityLabel props or label props to know that it is login form and then use runAfterInteractions like this
> 
> ```
> if (props.accessibilityLabel={translate('loginForm.phoneOrEmail')}) {
>             
>             InteractionManager.runAfterInteractions(() => {
>                 if (hasValueRef.current || isFocused || isInputAutoFilled(input.current)) {
>                     activateLabel();
>                 } else {
>                     deactivateLabel();
>                 }
>             });
>         }
> ```
> 
> What do you think about this way?

That's an intelligent idea. You think it's better than introducing new prop?

**Comment by situchan** [comment]
As this is tricky, I think we can fix only original issue without `InteractionManager.runAfterInteractions`. We can leave the 2nd issue I reported as is for now as it's edge case - user will not be likely to leave app before loading login screen after launch.
What do you think?
cc: @francoisl

**Comment by francoisl** [comment]
An alternative would be to introduce a new boolean prop such as `useInteractionManagerOnFocusChangeEffects` (terrible name but I can't think of something better right now), and then pass it on the inputs of the login page?

If that's not an option, then yes I'd say let's just fix the original issue without using `runAfterInteractions`. While the accessibility prop solution would work, it sounds like it would become messy to maintain if we end up making new pages that would require the same logic.

**Comment by situchan** [comment]
> If that's not an option, then yes I'd say let's just fix the original issue without using `runAfterInteractions`

I am 👍 on this.

**Comment by situchan** [comment]
For crash safe, let's add `!input ||` condition here:
https://github.com/Expensify/App/blob/6e055d073dff463097e5fed7cfed234bcfcde845/src/libs/isInputAutoFilled.js#L9

File: config/webpack/webpack.dev.js [hunk 3]
```diff
@@ -5,7 +5,7 @@ const {merge} = require('webpack-merge');
 const {TimeAnalyticsPlugin} = require('time-analytics-webpack-plugin');
 const getCommonConfig = require('./webpack.common');
 
-const BASE_PORT = 8082;
+const BASE_PORT = 8080;
```
Comment (situchan) [inline config/webpack/webpack.dev.js:8]: revert this change

**Review by situchan — APPROVED** [review]
LGTM :shipit: 
@francoisl all yours

**Review by francoisl — APPROVED** [review]

**Comment by francoisl** [comment]
> Please add a note explaining why this was done and remove the Emergency label if this is not an emergency.

Tests were passing, I think this is a bug with melvin.

**Comment by mvtglobally** [comment]
@dukenv0307 Can you update the steps? What link are we visiting? need more details to QA

**Comment by situchan** [comment]
> @dukenv0307 Can you update the steps? What link are we visiting? need more details to QA

It's just app start link
https://staging.new.expensify.com/
